sacred/utils.py

```diff
@@ -565,14 +565,18 @@ def filtered_traceback_format(tb_exception, chain=True):
             yield from filtered_traceback_format(tb_exception.__context__, chain=chain)
             yield tb._context_message
     yield "Traceback (most recent calls WITHOUT Sacred internals):\n"
-    current_tb = tb_exception.exc_traceback
-    while current_tb is not None:
-        if not _is_sacred_frame(current_tb.tb_frame):
-            stack = tb.StackSummary.extract(
-                tb.walk_tb(current_tb), limit=1, lookup_lines=True, capture_locals=False
-            )
-            yield from stack.format()
-        current_tb = current_tb.tb_next
+    if hasattr(tb_exception, "exec_traceback"):
+        current_tb = tb_exception.exc_traceback
+        while current_tb is not None:
+            if not _is_sacred_frame(current_tb.tb_frame):
+                stack = tb.StackSummary.extract(
+                    tb.walk_tb(current_tb),
+                    limit=1,
+                    lookup_lines=True,
+                    capture_locals=False,
+                )
+                yield from stack.format()
+            current_tb = current_tb.tb_next
```

Comment:
I'm not sure yet if this is the correct way of doing it. Is this a Python version specific issue? If so, then probably something in the API changed. If not, still something is wrong and it would be best to figure out what. I'm not completely sure, but can this hide the exception traceback completely?

Reply:
I had a look through the Python docu since 3.5 where `TracebackException` was introduced. `exc_traceback` is always needed in the constructor but it is never mentioned otherwise.

Reply:
Looking at the code (https://github.com/python/cpython/blob/117830de332c8dfbd9a437c0968e16e11aa7e6a1/Lib/traceback.py#L500, https://github.com/python/cpython/blob/3.8/Lib/traceback.py), they removed the `exc_traceback` in Python 3.8. So, it already paid off to update the tests to Python 3.8. The information is now only saved in the `TracebackException.stack` attribute. This attribute is mentioned in the docs, so it should be the way to go. It should contain all information, but I don't know yet how to filter the frames from the `StackSummary` object.

Reply:
This idea could work: 
```python
yield from tb.StackSummary.from_list(
    [sf for sf in tb_exception.stack if not sf.filename.startswith(sacred.__path__[0])]
).format()
```

Reply:
That seems to work. But I'm not entirely satisfied with the import statement.

Reply:
Hmm, then, it could work to use `__path__[0]` directly without importing sacred. You have to check if this really contains the correct path. It is important here that `__path__[0]` is a prefix of all paths to sacred files.

Another idea: subclass the `TracebackException` and use `_is_sacred_frame` while creating the stack summary. I did a quick check in a notebook, but this has the disadvantage of copying the whole `TracebackException.__init__`:
```python
def _is_sacred_frame(frame):
    return frame.f_globals["__name__"].split(".")[0] == "sacred"

def walk_tb_filtered(frame_gen):
    for tb_frame, tb_lineno in tb.walk_tb(frame_gen):
        if not _is_sacred_frame(tb_frame):
            yield tb_frame, tb_lineno

class FilteredTracebackException(tb.TracebackException):
    def __init__(self, exc_type, exc_value, exc_traceback, *, limit=None,
            lookup_lines=True, capture_locals=False, _seen=None):
        if _seen is None:
            _seen = set()
        _seen.add(id(exc_value))
        if (exc_value and exc_value.__cause__ is not None
            and id(exc_value.__cause__) not in _seen):
            cause = TracebackException(
                type(exc_value.__cause__),
                exc_value.__cause__,
                exc_value.__cause__.__traceback__,
                limit=limit,
                lookup_lines=False,
                capture_locals=capture_locals,
                _seen=_seen)
        else:
            cause = None
        if (exc_value and exc_value.__context__ is not None
            and id(exc_value.__context__) not in _seen):
            context = TracebackException(
                type(exc_value.__context__),
                exc_value.__context__,
                exc_value.__context__.__traceback__,
                limit=limit,
                lookup_lines=False,
                capture_locals=capture_locals,
                _seen=_seen)
        else:
            context = None
        self.exc_traceback = exc_traceback
        self.__cause__ = cause
        self.__context__ = context
        self.__suppress_context__ = \
            exc_value.__suppress_context__ if exc_value else False
        # This here filters the traceback
        self.stack = tb.StackSummary.extract(
            walk_tb_filtered(exc_traceback), limit=limit, lookup_lines=lookup_lines,
            capture_locals=capture_locals)
        self.exc_type = exc_type
        self._str = tb._some_str(exc_value)
        if exc_type and issubclass(exc_type, SyntaxError):
            # Handle SyntaxError's specially
            self.filename = exc_value.filename
            self.lineno = str(exc_value.lineno)
            self.text = exc_value.text
            self.offset = exc_value.offset
            self.msg = exc_value.msg
        if lookup_lines:
            self._load_lines()

try:
    c()
except:
    exc_type, exc_value, exc_traceback = sys.exc_info()
    tb_exception = FilteredTracebackException(
        exc_type, exc_value, exc_traceback, limit=None
    )
    
print(''.join(tb_exception.format()))
```
prints
```
Traceback (most recent call last):
  File "<ipython-input-25-f30411748fd8>", line 24, in <module>
    c()
  File "<ipython-input-4-ea5c93fb6d2f>", line 3, in c
    raise RuntimeError()
RuntimeError
```
Not sure yet if this is better than checking for the file name.

Reply:
Yet another not perfect solution:

Remove non-sacred frames directly from the traceback. This unfortunately doesn't adjust the back pointer in the frames (it's read-only), but it works for this purpose.
```python
def filter_traceback(traceback):
    first = prev = None
    while traceback is not None:
        if not _is_sacred_frame(traceback.tb_frame):
            if prev:
                prev.tb_next = traceback
            if first is None:
                prev = first = traceback
            prev = traceback
        traceback = traceback.tb_next
    return first

try:
    c()
except:
    exc_type, exc_value, exc_traceback = sys.exc_info()
    tb_exception = FilteredTracebackException(
        exc_type, exc_value, filter_traceback(exc_traceback), limit=None
    )
    
print(''.join(tb_exception.format()))
```

Reply:
I like the sub classing idea. Found a way to boild down __init__ to just the stuff needed.